pkg/actuators/awsplacementgroup/controller.go

```diff
@@ -3,91 +3,623 @@ package awsplacementgroup
+		// as placement group resources on AWS shouldn't be deleted if they are running in Unmanaged mode.
+		if controllerutil.ContainsFinalizer(awsPlacementGroup, awsPlacementGroupFinalizer) {
+			controllerutil.RemoveFinalizer(awsPlacementGroup, awsPlacementGroupFinalizer)
+			logger.Info("removing finalizer from aws placement group")
+
+			return ctrl.Result{Requeue: true}, nil
+		}
+
+		logger.Info("ignoring unmanaged aws placement group")
+		// Return and requeue the placement group even if it is Unmanaged to keep syncing up its ObservedConfiguration.
+		return ctrl.Result{RequeueAfter: requeueIn(awsPlacementGroup.Status.ExpiresAt.Time)}, nil
+	}
+
+	// If object DeletionTimestamp is zero, it means the object is not being deleted
+	// so clean up relevant resources.
+	if awsPlacementGroup.DeletionTimestamp.IsZero() {
+		if !controllerutil.ContainsFinalizer(awsPlacementGroup, awsPlacementGroupFinalizer) {
+			controllerutil.AddFinalizer(awsPlacementGroup, awsPlacementGroupFinalizer)
+			logger.Info("adding finalizer to aws placement group")
+
+			return ctrl.Result{Requeue: true}, nil
+		}
+	}
+
+	// If object DeletionTimestamp is not zero, it means the object is being deleted
+	// so clean up relevant resources.
+	if !awsPlacementGroup.DeletionTimestamp.IsZero() {
+		// No-op if the finalizer has been removed.
+		if !controllerutil.ContainsFinalizer(awsPlacementGroup, awsPlacementGroupFinalizer) {
+			logger.Info("reconciling aws placement group results in a no-op as there is no finalizer")
+			return ctrl.Result{}, nil
+		}
+
+		logger.Info("reconciling aws placement group triggers deletion")
+
+		if err := deletePlacementGroup(awsClient, logger, awsPlacementGroup, infra); err != nil {
+			werr := fmt.Errorf("failed to delete aws placement group: %w", err)
+			meta.SetStatusCondition(&awsPlacementGroup.Status.Conditions, metav1.Condition{
+				Type:    "Deleting",
+				Status:  metav1.ConditionTrue,
+				Message: werr.Error(),
+				Reason:  deletionFailedConditionReason,
+			})
+
+			return ctrl.Result{}, werr
+		}
+
+		// Remove finalizer on successful deletion.
+		controllerutil.RemoveFinalizer(awsPlacementGroup, awsPlacementGroupFinalizer)
+		logger.Info("removing finalizer after successful aws placement group deletion")
+
+		return ctrl.Result{}, nil
+	}
+
+	// Conditionally create or check the placement group.
+	if err := checkOrCreatePlacementGroup(awsClient, logger, awsPlacementGroup, infra); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{RequeueAfter: requeueIn(awsPlacementGroup.Status.ExpiresAt.Time)}, nil
+}
+
+// mergeInfrastructureAndAWSPlacementGroupSpecTags merge list of tags from AWSPlacementGroup provider spec and Infrastructure object platform spec.
+// Machine tags have precedence over Infrastructure.
+func mergeInfrastructureAndAWSPlacementGroupSpecTags(awsPlacementGroupSpecTags []machinev1beta1.TagSpecification, infra *configv1.Infrastructure) []machinev1beta1.TagSpecification {
```

Comment:
All of this tag management stuff is basically the same as for the Machine itself, wonder if we can deduplicate somehow, any thoughts? Could be a follow up PR

Reply:
Yes, that's a good point. I think a followup PR would be good. I'd prefer not to touch the machine controller in this one. I'll make a note to open a followup PR once this is merged.